x-pack/metricbeat/module/aws/ec2/ec2.go

```diff
@@ -168,14 +186,48 @@ func (m *MetricSet) createCloudWatchEvents(getMetricDataResults []cloudwatch.Met
 	// Find a timestamp for all metrics in output
 	timestamp := aws.FindTimestamp(getMetricDataResults)
 	if !timestamp.IsZero() {
+	NextOutput:
 		for _, output := range getMetricDataResults {
 			if len(output.Values) == 0 {
 				continue
 			}
+
 			exists, timestampIdx := aws.CheckTimestampInArray(timestamp, output.Timestamps)
 			if exists {
-				labels := strings.Split(*output.Label, " ")
+				labels := strings.Split(*output.Label, labelSeperator)
 				instanceID := labels[instanceIDIdx]
+
+				// Add tags
+				tags := instanceOutput[instanceID].Tags
+				var tagKeys []string
+				var tagValues []string
+				for _, tag := range tags {
+					tagKeys = append(tagKeys, *tag.Key)
+					tagValues = append(tagValues, *tag.Value)
+				}
+
+				if m.Tags == nil {
+					for _, tag := range tags {
+						events[instanceID].ModuleFields.Put("tags."+*tag.Key, *tag.Value)
+					}
+				} else {
+					// check with each tag filter
+					// If tag filter doesn't exist in tagKeys/tagValues, then do not
+					// store tags for this instance into event.
+					for filterIdx, tagFilter := range m.Tags {
+						if exists, idx := aws.StringInSlice(tagFilter.Name, tagKeys); !exists || tagValues[idx] != tagFilter.Value {
+							continue NextOutput
+						}
+
+						// If code runs here, means tag filters are all matched.
+						if filterIdx == len(m.Tags)-1 {
+							for _, tag := range tags {
+								events[instanceID].ModuleFields.Put("tags."+*tag.Key, *tag.Value)
+							}
+						}
+					}
+				}
```

Comment:
This method is starting to get complicated, maybe most of the added code could go to a separate function.

Also maybe the code for adding tags, and the code for filtering based on tags could be separated.